File: frontend/src/types/timeline.ts
/**
 * Mirrors `server/app/models/timeline.py`. Keep the two in sync.
 * All timestamps are ISO 8601 with an offset; they are converted to the
 * configured local timezone only when formatted for display.
 */

export type EventType = 'point' | 'interval' | 'continuous';
export type Origin = 'measured' | 'derived';
export type DataQuality = 'high' | 'medium' | 'low' | 'unknown';

export interface Provenance {
  rawRecordIds: string[];
  sourceEntityIds: string[];
  transformationRule?: string | null;
  ruleVersion?: string | null;
  thresholds: Record<string, unknown>;
  inputTimeRange?: string[] | null;
  outputTimestamp?: string | null;
  missingDataAssumptions: string[];
  notes: string[];
}

export interface TimelineEvent {
  id: string;
  phenotype: string;
  label: string;
  eventType: EventType;
  startTime: string;
  endTime?: string | null;
  value?: number | string | null;
  unit?: string | null;
  source: string;
  device?: string | null;
  entityId?: string | null;
  measuredOrDerived: Origin;
  confidence?: number | null;
  dataQuality: DataQuality;
  category?: string | null;
  continuesBefore: boolean;
  continuesAfter: boolean;
  metadata: Record<string, unknown>;
  provenance?: Provenance | null;
}

export interface SeriesPoint {
  timestamp: string;
  value: number;
  quality?: number | null;
}

export interface SeriesGap {
  startTime: string;
  endTime: string;
  reason?: string | null;
}

export interface TimelineSeries {
  id: string;
  phenotype: string;
  label: string;
  unit: string;
  source: string;
  device?: string | null;
  entityId?: string | null;
  measuredOrDerived: Origin;
  points: SeriesPoint[];
  gaps: SeriesGap[];
  minValue?: number | null;
  maxValue?: number | null;
  style: 'primary' | 'secondary';
  metadata: Record<string, unknown>;
  provenance?: Provenance | null;
}

export type AccentToken =
  | 'green'
  | 'blue'
  | 'indigo'
  | 'purple'
  | 'orange'
  | 'teal'
  | 'sky'
  | 'cyan';

export interface Lane {
  id: string;
  phenotype: string;
  label: string;
  description: string;
  accent: AccentToken;
  available: boolean;
  unavailableReason?: string | null;
  units: string[];
  events: TimelineEvent[];
  series: TimelineSeries[];
  sources: string[];
}

export interface CoverageWindow {
  startTime: string;
  endTime: string;
  label: string;
}

export interface DayCoverage {
  overallFraction: number;
  perLane: Record<string, number>;
  missingPeriods: CoverageWindow[];
}

export interface SyncSummary {
  dateProcessed: string;
  localTimezone: string;
  dayStart: string;
  dayEnd: string;
  dayLengthHours: number;
  sourcesChecked: string[];
  rawRecordCount: number;
  normalizedEventCount: number;
  derivedFeatureCount: number;
  seriesPointCount: number;
  coverage: DayCoverage;
  warnings: string[];
  errors: string[];
  startedAt?: string | null;
  completedAt?: string | null;
}

export interface DayTimeline {
  date: string;
  localTimezone: string;
  dayStart: string;
  dayEnd: string;
  dayLengthHours: number;
  generatedAt: string;
  lanes: Lane[];
  summary: SyncSummary;
  highlights: string[];
  mockData: boolean;
}

export type SourceStatus =
  | 'connected'
  | 'disconnected'
  | 'syncing'
  | 'error'
  | 'mock_data';

export type SourceTransport = 'mcp' | 'rest' | 'mock' | 'file';

export interface DataSource {
  id: string;
  name: string;
  status: SourceStatus;
  /** The MCP server this source corresponds to, as named in the MCP client. */
  mcpServer?: string | null;
  transport: SourceTransport;
  provider?: string | null;
  capabilities: string[];
  detail?: string | null;
  lastSync?: string | null;
  entityCount?: number | null;
  /** False when the source answered but had nothing for the displayed day. */
  hasData: boolean;
  /** False when switched off, in which case it was never contacted. */
  selected: boolean;
  /** Position in the merge order; lower wins when two sources share a metric. */
  priority: number | null;
}

export interface DataSourceReport {
  sources: DataSource[];
  mockData: boolean;
  checkedAt?: string | null;
}

export interface EventDetailsResponse {
  event: TimelineEvent;
  laneId: string;
  date: string;
  rawRecordCount: number;
  rawRecords: RawRecordSummary[];
}

export interface RawRecordSummary {
  id: string;
  source: string;
  stream: string;
  entityId?: string | null;
  device?: string | null;
  timestamp: string;
  endTimestamp?: string | null;
  value?: number | string | null;
  unit?: string | null;
}

/** A single sample the user clicked on a continuous line. */
export interface SeriesSelection {
  kind: 'series-point';
  laneId: string;
  series: TimelineSeries;
  point: SeriesPoint;
}

export interface EventSelection {
  kind: 'event';
  laneId: string;
  event: TimelineEvent;
}

export type Selection = EventSelection | SeriesSelection;

export function selectionKey(selection: Selection): string {
  return selection.kind === 'event'
    ? `event:${selection.event.id}`
    : `point:${selection.series.id}:${selection.point.timestamp}`;
}

File: server/app/models/timeline.py
"""The shared internal event schema.

Everything downstream of the connectors speaks these types: normalization emits
them, feature engineering annotates them, the API serialises them, and the
frontend mirrors them in `frontend/src/types/timeline.ts`.

All timestamps are timezone-aware and serialised as ISO 8601. Conversion to a
human-readable local time happens only in the presentation layer.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

EventType = Literal["point", "interval", "continuous"]
Origin = Literal["measured", "derived"]

Phenotype = Literal[
    "activity",
    "heart_rate",
    "hrv",
    "readiness",
    "sleep",
    "temperature",
    "environment",
    "presence",
    "location",
]


class CamelModel(BaseModel):
    """Serialises to camelCase for the TypeScript client, accepts either form."""

    model_config = ConfigDict(
        populate_by_name=True,
        alias_generator=lambda field: "".join(
            part if index == 0 else part.capitalize()
            for index, part in enumerate(field.split("_"))
        ),
    )


class Provenance(CamelModel):
    """Why a derived feature exists, and what it was derived from."""

    raw_record_ids: list[str] = Field(default_factory=list)
    source_entity_ids: list[str] = Field(default_factory=list)
    transformation_rule: str | None = None
    rule_version: str | None = None
    thresholds: dict[str, Any] = Field(default_factory=dict)
    input_time_range: list[str] | None = None
    output_timestamp: str | None = None
    missing_data_assumptions: list[str] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class TimelineEvent(CamelModel):
    """A point or interval that occupies a lane."""

    id: str
    phenotype: str
    label: str
    event_type: EventType
    start_time: datetime
    end_time: datetime | None = None
    value: float | str | None = None
    unit: str | None = None
    source: str
    device: str | None = None
    entity_id: str | None = None
    measured_or_derived: Origin
    confidence: float | None = None
    data_quality: Literal["high", "medium", "low", "unknown"] = "unknown"
    category: str | None = None
    """Sub-kind used by the renderer, e.g. `workout`, `nap`, `light_bright`."""
    continues_before: bool = False
    continues_after: bool = False
    metadata: dict[str, Any] = Field(default_factory=dict)
    provenance: Provenance | None = None

    @property
    def duration_seconds(self) -> float | None:
        if self.end_time is None:
            return None
        return (self.end_time - self.start_time).total_seconds()


class SeriesPoint(CamelModel):
    timestamp: datetime
    value: float
    quality: float | None = None


class SeriesGap(CamelModel):
    """An interval where a continuous stream had no samples."""

    start_time: datetime
    end_time: datetime
    reason: str | None = None


class TimelineSeries(CamelModel):
    """A continuous measurement rendered as a line."""

    id: str
    phenotype: str
    label: str
    unit: str
    source: str
    device: str | None = None
    entity_id: str | None = None
    measured_or_derived: Origin = "measured"
    points: list[SeriesPoint] = Field(default_factory=list)
    gaps: list[SeriesGap] = Field(default_factory=list)
    min_value: float | None = None
    max_value: float | None = None
    style: Literal["primary", "secondary"] = "primary"
    metadata: dict[str, Any] = Field(default_factory=dict)
    provenance: Provenance | None = None


class Lane(CamelModel):
    """One horizontal swimlane."""

    id: str
    phenotype: str
    label: str
    description: str
    accent: str
    """Semantic colour token resolved by the frontend, e.g. `green`."""
    available: bool = True
    unavailable_reason: str | None = None
    units: list[str] = Field(default_factory=list)
    events: list[TimelineEvent] = Field(default_factory=list)
    series: list[TimelineSeries] = Field(default_factory=list)
    sources: list[str] = Field(default_factory=list)


class CoverageWindow(CamelModel):
    start_time: datetime
    end_time: datetime
    label: str


class DayCoverage(CamelModel):
    """How much of the day each stream actually observed."""

    overall_fraction: float = 0.0
    per_lane: dict[str, float] = Field(default_factory=dict)
    missing_periods: list[CoverageWindow] = Field(default_factory=list)


class SyncSummary(CamelModel):
    date_processed: str
    local_timezone: str
    day_start: datetime
    day_end: datetime
    day_length_hours: float
    sources_checked: list[str] = Field(default_factory=list)
    raw_record_count: int = 0
    normalized_event_count: int = 0
    derived_feature_count: int = 0
    series_point_count: int = 0
    coverage: DayCoverage = Field(default_factory=DayCoverage)
    warnings: list[str] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)
    started_at: datetime | None = None
    completed_at: datetime | None = None


class DayTimeline(CamelModel):
    """The payload the Yesterday page renders."""

    date: str
    local_timezone: str
    day_start: datetime
    day_end: datetime
    day_length_hours: float
    generated_at: datetime
    lanes: list[Lane] = Field(default_factory=list)
    summary: SyncSummary
    highlights: list[str] = Field(default_factory=list)
    """Neutral temporal descriptions. Never causal claims."""
    mock_data: bool = False
